### PR TITLE
Docked composer

### PR DESCRIPTION
**Pull request by 2FALockedMeOutOfMyAcct**
### Additional description
Adds [Docked Composer](https://www.lightningdesignsystem.com/components/docked-composer/) to DSR
Screen Shot 2022-08-26 at 10 29 48 AM

---

### CONTRIBUTOR checklist (do not remove)
Please complete for every pull request

* [x] First-time contributors should sign the Contributor License Agreement. It's a fancy way of saying that you are giving away your contribution to this project. If you haven't before, wait a few minutes and a bot will comment on this pull request with instructions.
* [x] `npm run lint:fix` has been run and linting passes.
* [x] Mocha, Jest (Storyshots), and `components/component-docs.json` CI tests pass (`npm test`).
* [x] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [x] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [x] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [x] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).

### REVIEWER checklist (do not remove)

* [ ] CircleCI tests pass. This includes linting, Mocha, Jest, Storyshots, and `components/component-docs.json` tests.
* [ ] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [ ] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [ ] The Accessibility panel of each Storybook story has 0 violations (aXe). Open [http://localhost:9001/](http://localhost:9001/).
* [ ] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).
###### Required only if there are markup / UX changes
* [ ] Add year-first date and commit SHA to `last-slds-markup-review` in `package.json` and push.
* [ ] Request a review of the deployed Heroku app by the Salesforce UX Accessibility Team.
* [ ] Add year-first review date, and commit SHA, `last-accessibility-review`, to `package.json` and push.
* [ ] While the contributor's branch is checked out, run `npm run local-update` within locally cloned [site repo](https://github.com/salesforce-ux/design-system-react-site) to confirm the site will function correctly at the next release.

### CODE REVIEW

File: components/docked-composer/index.jsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+									title={this.props.header}
+								>
+									{this.props.header}
+								</h2>
+							</div>
+						</div>
+						<div className="slds-col_bump-left slds-shrink-none">
+							<Button
+								id={`${this.getId()}-minimize-button`}
+								title="Minimize"
```
Comment (meizibupt) [inline components/docked-composer/index.jsx:85]: should we localize titles?

File: components/docked-composer/index.jsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	/**
+	 * Whether the docked composer is open or closed.
+	 * _Tested with snapshot testing._
+	 */
+	isOpen: PropTypes.bool,
+	/**
+	 * Text to display in the header.
+	 * _Tested with snapshot testing._
+	 */
+	header: PropTypes.string,
```
Comment (meizibupt) [inline components/docked-composer/index.jsx:22]: should we allow user to add a node for header too?

File: components/docked-composer/index.jsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import shortid from 'shortid';
+
+import Button from '../button';
+
+const propTypes = {
+	/**
```
Comment (meizibupt) [inline components/docked-composer/index.jsx:8]: we can also add a css class prop to allow overriding  css

File: components/docked-composer/__examples__/base.jsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import React from 'react';
+
+import DockedComposer from '~/components/docked-composer';
+import Accordion from '~/components/accordion/__examples__/base';
+
+class Example extends React.Component {
+	static displayName = 'DockedComposerExample';
+
+	state = {
+		isOpen: true,
```
Comment (alexxpan) [inline components/docked-composer/__examples__/base.jsx:11]: This naming convention is a bit confusing, do you think `isExpanded` and `isOpen` would be a bit less ambiguous?

**Comment by 2FALockedMeOutOfMyAcct** [comment]
Minimized and expanded states:
Screen Shot 2022-08-29 at 12 14 29 PM
Screen Shot 2022-08-29 at 12 14 43 PM
3 votes in favor of this behavior: @gfyhui , @alexxpan and Jonathon

**Review by meizibupt — DISMISSED** [review]

**Review by alexxpan — DISMISSED** [review]

**Review by meizibupt — APPROVED** [review]

**Review by garygong — DISMISSED** [review]

File: components/docked-composer/__tests__/docked-composer.browser-test.jsx [hunk 5]
```diff
@@ -0,0 +1,94 @@
+// Import your external dependencies
+import React from 'react';
+
+import chai, { expect } from 'chai';
+import chaiEnzyme from 'chai-enzyme';
+import { mount } from 'enzyme';
+import IconSettings from '../../icon-settings';
+
+// Import your internal dependencies (for example):
```
Comment (alexzherdev) [inline components/docked-composer/__tests__/docked-composer.browser-test.jsx:9]: These comments seem like they can be removed

File: components/docked-composer/__examples__/minimized.jsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+				body={<Accordion />}
+				classNameContainer="hello-world"
+				assistiveText={{
+					expandButton: 'ExPaNd',
+					minimizeButton: 'MiNiMiZe',
+					closeButton: 'ClOsE',
+				}}
+				id="docked-composer"
+				events={{
+					onMinimize: this.handleMinimize,
```
Comment (interactivellama) [inline components/docked-composer/__examples__/minimized.jsx:40]: I haven't dove deep into this yet, but these look like on request pre-state change actions.
https://github.com/salesforce/design-system-react/blob/master/docs/codebase-overview.md#prefix-onrequestaction-to-pre-state-change-callbacks-functions

File: components/docked-composer/index.jsx [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+				className={`slds-docked_container${
+					this.props.classNameContainer
+						? ` ${this.props.classNameContainer}`
+						: ''
+				}`}
+			>
+				<section
+					className={`slds-docked-composer slds-grid slds-grid_vertical ${sectionClassName}`}
+					role="dialog"
+					aria-labelledby="dialog-heading-id-1"
```
Comment (interactivellama) [inline components/docked-composer/index.jsx:89]: All ids need to be prefixed by an id or a random id if not set `this.generatedId`

File: components/docked-composer/index.jsx [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+								>
+									{this.props.header}
+								</h2>
+							</div>
+						</div>
+						<div className="slds-col_bump-left slds-shrink-none">
+							{this.props.isOpen ? (
+								<Button
+									id={`${this.getId()}-minimize-button`}
+									title={this.props.assistiveText?.minimizeButton || 'Minimize'}
```
Comment (interactivellama) [inline components/docked-composer/index.jsx:111]: I think was would be typically done by a default props object, so no one has to search render functions for text literals.

File: components/docked-composer/index.jsx [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+									assistiveText={{
+										icon:
+											this.props.assistiveText?.minimizeButton || 'Minimize',
```
Comment (interactivellama) [inline components/docked-composer/index.jsx:114]: Same here.

File: components/docked-composer/index.jsx [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+									onClick={this.props.events.onMinimize}
+									iconCategory="utility"
+									iconName="minimize_window"
+									iconVariant="bare"
+									variant="icon"
+								/>
+							) : (
+								<Button
+									id={`${this.getId()}-expand-button`}
+									title={this.props.assistiveText?.expandButton || 'Expand'}
```
Comment (interactivellama) [inline components/docked-composer/index.jsx:125]: Same.

File: components/docked-composer/index.jsx [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+									assistiveText={{
+										icon: this.props.assistiveText?.expandButton || 'Expand',
```
Comment (interactivellama) [inline components/docked-composer/index.jsx:127]: Same.

**Review by interactivellama — CHANGES_REQUESTED** [review]
First, this is a great pull request! I did not pull this down to manually test it, but did left a few minor request and referenced the overview doc. Thank you for this great contribution!

**Review by interactivellama — APPROVED** [review]

**Comment by interactivellama** [comment]
@alexzherdev Merge when you ready to.

**Comment by alexzherdev** [comment]
Not super important for this simple component, but I wonder if we should encourage new components to be written as functions. At this point I don't think there are any pre-16.8 consumers?

**Comment by interactivellama** [comment]
@alexzherdev I'm cool with that. Please update the docs when convenient. I believe DataTable has hooks in it.

**Comment by interactivellama** [comment]
Also, love the account name: @2FALockedMeOutOfMyAcct

**Comment by 2FALockedMeOutOfMyAcct** [comment]
Thanks for the reviews and merging. When will the next DSR release be @interactivellama @alexzherdev ?

**Comment by alexzherdev** [comment]
@interactivellama is the automated release build still broken?

**Comment by interactivellama** [comment]
Yes.

I could use some help in triaging the authentication, since the security protections went up.

**Comment by interactivellama** [comment]
@2FALockedMeOutOfMyAcct Please open a release PR and I will pull down and push up a tag.

**Comment by alexzherdev** [comment]
@interactivellama I opened #3063 if that's what you meant?

**Comment by interactivellama** [comment]
I'll try to get this out by End of week.

**Comment by interactivellama** [comment]
https://github.com/salesforce/design-system-react/releases/tag/v0.10.49

**Comment by interactivellama** [comment]
There are additional issues with the NPM publishing process. I'm in talks with Jim in OSS group on getting different permissions set up. The tag is present in Github but we do not have a NPM package up today.

**Pull request merged**